File: sorting/heap_sort.py
def max_heapify(arr, n, i):
    largest = i
    l = 2 *i + 1
    r = 2 * i + 2

    if l < n and arr[l] > arr[largest]:
        largest = l
    
    if r < n and arr[r] > arr[largest]:
        largest = r

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i] 
        max_heapify(arr, n, largest)  

def Build_Max_Heap(arr,n):
    for i in range(n // 2 - 1, -1, -1):
        max_heapify(arr, n, i)

def heap_sort(arr,n):
    Build_Max_Heap(arr,n)
    
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  
        max_heapify(arr, i, 0)  


arr = list(map(int, input("Enter numbers separated by space: ").split()))
print("The array of list before sorting: ", arr)
heap_sort(arr,len(arr))
print("Sorted array:", arr)

File: sorting/Heap_sort.c
#include <stdio.h>

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

void maxHeapify(int A[], int n, int i)
{
    int largest = i;
    int l = 2 * i;
    int r = 2 * i + 1;

    if (l <= n && A[l] > A[largest])
        largest = l;
    if (r <= n && A[r] > A[largest])
        largest = r;

    if (largest != i)
    {
        swap(&A[i], &A[largest]);
        maxHeapify(A, n, largest);
    }
}

void heapSort(int A[], int n)
{

    for (int i = n / 2; i >= 1; i--)
    {
        maxHeapify(A, n, i);
    }

    for (int i = n; i >= 2; i--)
    {
        swap(&A[1], &A[i]);
        maxHeapify(A, i - 1, 1);
    }
}

int main()
{
    int A[] = {0, 4, 10, 3, 5, 1}; 
    int n = 5;

    heapSort(A, n);

    printf("Sorted array: ");
    for (int i = 1; i <= n; i++)
        printf("%d ", A[i]);
    printf("\n");

    return 0;
}
